File: backend/src/routes/history.rs
use axum::{
    extract::State,
    routing::{get, post, delete},
    Json, Router,
};
use serde::{Deserialize, Serialize};

use crate::auth::OptionalAuthUser;
use crate::db::{self, DbPool, WatchHistoryItem};

#[derive(Deserialize)]
pub struct AddHistoryRequest {
    video_id: String,
    title: String,
    thumbnail: String,
}

#[derive(Serialize)]
pub struct HistoryResponse {
    history: Vec<WatchHistoryItem>,
}

/// Add to watch history
async fn add_history(
    State(pool): State<DbPool>,
    auth: OptionalAuthUser,
    Json(req): Json<AddHistoryRequest>,
) -> Result<Json<serde_json::Value>, (axum::http::StatusCode, String)> {
    let user_id = auth.user_id_or_default().to_string();

    db::add_watch_history(&pool, &user_id, &req.video_id, &req.title, &req.thumbnail)
        .map_err(|e| (axum::http::StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "success": true })))
}

/// Get watch history
async fn get_history(
    State(pool): State<DbPool>,
    auth: OptionalAuthUser,
) -> Result<Json<HistoryResponse>, (axum::http::StatusCode, String)> {
    let user_id = auth.user_id_or_default().to_string();

    let history = db::get_watch_history(&pool, &user_id, 20)
        .map_err(|e| (axum::http::StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(HistoryResponse { history }))
}

#[derive(Deserialize)]
pub struct DeleteHistoryRequest {
    video_id: String,
}

/// Delete from watch history
async fn delete_history(
    State(pool): State<DbPool>,
    auth: OptionalAuthUser,
    Json(req): Json<DeleteHistoryRequest>,
) -> Result<Json<serde_json::Value>, (axum::http::StatusCode, String)> {
    let user_id = auth.user_id_or_default().to_string();

    db::delete_watch_history(&pool, &user_id, &req.video_id)
        .map_err(|e| (axum::http::StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "success": true })))
}

/// Clear all watch history
async fn clear_history(
    State(pool): State<DbPool>,
    auth: OptionalAuthUser,
) -> Result<Json<serde_json::Value>, (axum::http::StatusCode, String)> {
    let user_id = auth.user_id_or_default().to_string();

    db::clear_watch_history(&pool, &user_id)
        .map_err(|e| (axum::http::StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(serde_json::json!({ "success": true })))
}

pub fn routes(pool: DbPool) -> Router {
    Router::new()
        .route("/", get(get_history).post(add_history))
        .route("/delete", post(delete_history))
        .route("/clear", post(clear_history))
        .with_state(pool)
}

File: worker/src/routes/history.ts
import { Hono } from 'hono'
import type { Env } from '../env'
import { getOptionalAuth } from '../middleware/auth'
import * as db from '../db'

export function historyRoutes() {
  const app = new Hono<{ Bindings: Env }>()

  // ─── Add watch history ───
  app.post('/add', async (c) => {
    const { user_id } = await getOptionalAuth(c)
    const body = await c.req.json<{
      video_id: string
      title: string
      thumbnail?: string
    }>()

    try {
      await db.addWatchHistory(
        c.env.DB,
        user_id,
        body.video_id,
        body.title,
        body.thumbnail ?? '',
      )
      return c.json({ success: true, data: null })
    } catch (e) {
      return c.json({ success: false, error: `Failed to add history: ${e}` })
    }
  })

  // ─── Get watch history ───
  app.get('/list', async (c) => {
    const { user_id } = await getOptionalAuth(c)
    const limit = parseInt(c.req.query('limit') || '50')

    try {
      const history = await db.getWatchHistory(c.env.DB, user_id, limit)
      return c.json({ success: true, data: { history, total: history.length } })
    } catch (e) {
      return c.json({ success: false, error: `Failed to get history: ${e}` })
    }
  })

  // ─── Delete single history entry ───
  app.delete('/delete/:videoId', async (c) => {
    const { user_id } = await getOptionalAuth(c)
    const videoId = c.req.param('videoId')

    try {
      await db.deleteWatchHistory(c.env.DB, user_id, videoId)
      return c.json({ success: true, data: null })
    } catch (e) {
      return c.json({ success: false, error: `Failed to delete history: ${e}` })
    }
  })

  // ─── Clear all watch history ───
  app.delete('/clear', async (c) => {
    const { user_id } = await getOptionalAuth(c)

    try {
      await db.clearWatchHistory(c.env.DB, user_id)
      return c.json({ success: true, data: null })
    } catch (e) {
      return c.json({ success: false, error: `Failed to clear history: ${e}` })
    }
  })

  return app
}
